File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.SmartReferenceProxyExample/RealDatabaseConnection.cs
namespace ProxyDesignPattern.SmartReferenceProxyExample;

/// <summary>
/// RealSubject - Actual database connection (expensive resource).
/// </summary>
public class RealDatabaseConnection : IDatabaseConnection
{
    private readonly string _connectionId;
    private bool _isClosed;

    public RealDatabaseConnection()
    {
        _connectionId = Guid.NewGuid().ToString()[..8].ToUpper();
        Console.WriteLine($"[Connection-{_connectionId}] Database connection opened");
    }

    public void ExecuteQuery(string query)
    {
        if (_isClosed)
            throw new InvalidOperationException("Connection is closed");

        Console.WriteLine($"[Connection-{_connectionId}] Executing: {query}");
        Thread.Sleep(100); // Simulate query execution
        Console.WriteLine($"[Connection-{_connectionId}] Query completed");
    }

    public void Close()
    {
        if (!_isClosed)
        {
            _isClosed = true;
            Console.WriteLine($"[Connection-{_connectionId}] Connection closed");
        }
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.SmartReferenceProxyExample/src/RealDatabaseConnection.ts
import { IDatabaseConnection } from "./IDatabaseConnection";

/**
 * RealSubject - Actual database connection (expensive resource).
 */
export class RealDatabaseConnection implements IDatabaseConnection {
  private readonly _connectionId: string;
  private _isClosed = false;

  constructor() {
    this._connectionId = Math.random().toString(36).substring(2, 10).toUpperCase();
    console.log(`[Connection-${this._connectionId}] Database connection opened`);
  }

  executeQuery(query: string): void {
    if (this._isClosed) {
      throw new Error("Connection is closed");
    }

    console.log(`[Connection-${this._connectionId}] Executing: ${query}`);
    // Simulate query execution (synchronous for simplicity)
    console.log(`[Connection-${this._connectionId}] Query completed`);
  }

  close(): void {
    if (!this._isClosed) {
      this._isClosed = true;
      console.log(`[Connection-${this._connectionId}] Connection closed`);
    }
  }
}
